File: example_client.js
/**
 * Example client showing how to use the Vector Store Service from any Node.js/JavaScript application
 */

import fetch from 'node-fetch';
import fs from 'fs';

class VectorStoreClient {
    constructor(serviceUrl = "http://localhost:5001") {
        this.serviceUrl = serviceUrl;
    }

    async checkHealth() {
        try {
            const response = await fetch(`${this.serviceUrl}/health`);
            return await response.json();
        } catch (error) {
            return { status: "error", message: error.message };
        }
    }

    async getServiceInfo() {
        try {
            const response = await fetch(`${this.serviceUrl}/info`);
            return await response.json();
        } catch (error) {
            return { error: error.message };
        }
    }

    async uploadAndAttach(filePath, vectorStoreId = null, attributes = {}) {
        const payload = {
            file_path: filePath,
            attributes: attributes
        };

        if (vectorStoreId) {
            payload.vector_store_id = vectorStoreId;
        }

        try {
            const response = await fetch(`${this.serviceUrl}/upload_and_attach`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(payload)
            });
            return await response.json();
        } catch (error) {
            return { success: false, error: error.message };
        }
    }
}

async function exampleUsage() {
    console.log('🔧 Vector Store Service Client Example');
    console.log('='.repeat(50));

    // Initialize client
    const client = new VectorStoreClient("http://localhost:5001");

    // Check service health
    console.log('\n1. Checking service health...');
    const health = await client.checkHealth();
    console.log(`   Status: ${health.status || 'unknown'}`);

    if (health.status !== 'healthy') {
        console.log('   ❌ Service not available. Please start the service first.');
        return;
    }

    // Get service info
    console.log('\n2. Getting service information...');
    const info = await client.getServiceInfo();
    console.log(`   Service: ${info.service || 'unknown'}`);
    console.log(`   Version: ${info.version || 'unknown'}`);

    // Example file upload (you'll need to provide a real file path)
    const exampleFile = "/path/to/your/document.pdf";  // Change this to a real file

    if (fs.existsSync(exampleFile)) {
        console.log(`\n3. Uploading and attaching file: ${exampleFile}`);

        // Custom attributes for your application
        const customAttrs = {
            application: "my_web_app",
            user_id: "user_456",
            uploaded_by: "javascript_client",
            category: "test_document",
            priority: "high",
            department: "engineering"
        };

        const result = await client.uploadAndAttach(
            exampleFile,
            null, // Use default vector store
            customAttrs
        );

        if (result.success) {
            console.log('   ✅ Upload successful!');
            console.log(`   📄 File ID: ${result.file.id}`);
            console.log(`   📊 Size: ${result.file.bytes} bytes`);
            console.log(`   🏷️  Attributes applied: ${Object.keys(result.applied_attributes).length}`);
            console.log('\n   Auto-extracted metadata:');
            const attrs = result.applied_attributes;
            console.log(`      • Company: ${attrs.company || 'N/A'}`);
            console.log(`      • Document Type: ${attrs.document_type || 'N/A'}`);
            console.log(`      • Year: ${attrs.year || 'N/A'}`);
        } else {
            console.log(`   ❌ Upload failed: ${result.error}`);
        }
    } else {
        console.log(`\n3. Skipping file upload - file not found: ${exampleFile}`);
        console.log('   💡 Update the "exampleFile" path to test file upload');
    }

    console.log('\n' + '='.repeat(50));
    console.log('✅ Example completed. The service can be used from any application!');
}

// Export for use in other modules
export { VectorStoreClient };

// Run example if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
    exampleUsage().catch(console.error);
}

File: example_client.py
#!/usr/bin/env python3
"""
Example client showing how to use the Vector Store Service from any Python application
"""

import requests
import json
import os

class VectorStoreClient:
    def __init__(self, service_url="http://localhost:5001"):
        self.service_url = service_url
        
    def check_health(self):
        """Check if the service is healthy"""
        try:
            response = requests.get(f"{self.service_url}/health")
            return response.json()
        except Exception as e:
            return {"status": "error", "message": str(e)}
    
    def get_service_info(self):
        """Get service information and capabilities"""
        try:
            response = requests.get(f"{self.service_url}/info")
            return response.json()
        except Exception as e:
            return {"error": str(e)}
    
    def upload_and_attach(self, file_path, vector_store_id=None, attributes=None):
        """
        Upload file and attach to vector store with custom attributes
        
        Args:
            file_path (str): Path to the file to upload
            vector_store_id (str, optional): Vector store ID to use
            attributes (dict, optional): Custom attributes to add
            
        Returns:
            dict: Service response
        """
        payload = {
            "file_path": file_path,
            "attributes": attributes or {}
        }
        
        if vector_store_id:
            payload["vector_store_id"] = vector_store_id
            
        try:
            response = requests.post(
                f"{self.service_url}/upload_and_attach",
                headers={"Content-Type": "application/json"},
                json=payload
            )
            return response.json()
        except Exception as e:
            return {"success": False, "error": str(e)}

def example_usage():
    """Example of how to use the Vector Store Service from any application"""
    
    print("🔧 Vector Store Service Client Example")
    print("=" * 50)
    
    # Initialize client
    client = VectorStoreClient("http://localhost:5001")
    
    # Check service health
    print("\n1. Checking service health...")
    health = client.check_health()
    print(f"   Status: {health.get('status', 'unknown')}")
    
    if health.get('status') != 'healthy':
        print("   ❌ Service not available. Please start the service first.")
        return
    
    # Get service info
    print("\n2. Getting service information...")
    info = client.get_service_info()
    print(f"   Service: {info.get('service', 'unknown')}")
    print(f"   Version: {info.get('version', 'unknown')}")
    
    # Example file upload (you'll need to provide a real file path)
    example_file = "/path/to/your/document.pdf"  # Change this to a real file
    
    if os.path.exists(example_file):
        print(f"\n3. Uploading and attaching file: {example_file}")
        
        # Custom attributes for your application
        custom_attrs = {
            "application": "my_document_manager",
            "user_id": "user_123",
            "uploaded_by": "example_script",
            "category": "test_document",
            "priority": "normal",
            "department": "research"
        }
        
        result = client.upload_and_attach(
            file_path=example_file,
            attributes=custom_attrs
        )
        
        if result.get("success"):
            print("   ✅ Upload successful!")
            print(f"   📄 File ID: {result['file']['id']}")
            print(f"   📊 Size: {result['file']['bytes']} bytes")
            print(f"   🏷️  Attributes applied: {len(result['applied_attributes'])}")
            print("\n   Auto-extracted metadata:")
            attrs = result['applied_attributes']
            print(f"      • Company: {attrs.get('company', 'N/A')}")
            print(f"      • Document Type: {attrs.get('document_type', 'N/A')}")
            print(f"      • Year: {attrs.get('year', 'N/A')}")
        else:
            print(f"   ❌ Upload failed: {result.get('error')}")
    else:
        print(f"\n3. Skipping file upload - file not found: {example_file}")
        print("   💡 Update the 'example_file' path to test file upload")
    
    print("\n" + "=" * 50)
    print("✅ Example completed. The service can be used from any application!")

if __name__ == "__main__":
    example_usage()